### PR TITLE
fix(implicit-tiling): handle child tiles

### PR DESCRIPTION
**Pull request by dryabinin-actionengine**
In 3DTiles, not only the root tile can be implicit, but also the child tile. We have to deal with this as well.

### CODE REVIEW

File: modules/tile-converter/src/i3s-converter/helpers/gltf-attributes.ts [hunk 1]
```diff
@@ -49,35 +49,48 @@ export function prepareDataForAttributesConversion(tileContent: B3DMContent): B3
...
     gltfMaterials,
-    nodes: prepearedNodes,
+    nodes,
     cartographicOrigin,
     cartesianModelMatrix
   };
 }
+
+/**
+ * Traverse all nodes to replace all sensible data with copy.
```
Comment (ibgreen) [inline modules/tile-converter/src/i3s-converter/helpers/gltf-attributes.ts:66]: Can we improve / expand this comment. What does "sensible data" mean? Why do we replace it with copies? Does this increase memory pressure?

File: modules/tile-converter/src/i3s-attributes-worker.ts [hunk 2]
```diff
@@ -15,7 +15,7 @@ export type I3SAttributesWorkerOptions = {
 
 export type B3DMAttributesData = {
   gltfMaterials?: {id: string}[];
-  nodes: any;
+  nodes: any[];
```
Comment (ibgreen) [inline modules/tile-converter/src/i3s-attributes-worker.ts:18]: Any chance we can improve this typing? Or make it unknown?

File: modules/3d-tiles/src/lib/parsers/helpers/parse-3d-tile-subtree.ts [hunk 3]
```diff
@@ -11,7 +11,7 @@ const SUBTREE_FILE_VERSION = 1;
  * @returns
  */
 // eslint-disable-next-line max-statements
-export default async function parse3DTilesSubtree(data: ArrayBuffer): Promise<Subtree> {
+export default async function parse3DTilesSubtree(data: ArrayBuffer, options): Promise<Subtree> {
```
Comment (ibgreen) [inline modules/3d-tiles/src/lib/parsers/helpers/parse-3d-tile-subtree.ts:14]: Can we type the `options` parameter?

File: modules/3d-tiles/src/lib/parsers/helpers/parse-3d-tile-subtree.ts [hunk 4]
```diff
@@ -42,29 +42,49 @@ export default async function parse3DTilesSubtree(data: ArrayBuffer): Promise<Su
...
-      internalBinaryBuffer
+      internalBinaryBuffer,
+      options.basePath
     );
   }
 
   return subtree;
 }
 
+function resolveUri(uri, basePath) {
```
Comment (ibgreen) [inline modules/3d-tiles/src/lib/parsers/helpers/parse-3d-tile-subtree.ts:71]: Parameter types and return values, JSDoc

File: modules/3d-tiles/src/lib/parsers/helpers/parse-3d-tile-subtree.ts [hunk 5]
```diff
@@ -74,15 +94,17 @@ export default async function parse3DTilesSubtree(data: ArrayBuffer): Promise<Su
...
 ): Promise<ExplicitBitstream> {
   const bufferViewIndex = subtree[name].bufferView;
   const bufferView = subtree.bufferViews[bufferViewIndex];
   const buffer = subtree.buffers[bufferView.buffer];
 
   // External bitstream loading
   if (buffer.uri) {
-    const response = await fetchFile(buffer.uri);
+    const bufferUri = resolveUri(buffer.uri, basePath);
+    const response = await fetchFile(bufferUri);
```
Comment (ibgreen) [inline modules/3d-tiles/src/lib/parsers/helpers/parse-3d-tile-subtree.ts:107]: Should we be using `context.fetch` to make sure we don't get problems in workers etc? 

I am not 100% sure, but then we might not need to pass in basePath as I think it is already on context.

File: modules/3d-tiles/src/lib/parsers/parse-3d-tile-header.ts [hunk 6]
```diff
@@ -70,18 +70,35 @@ export function normalizeTileData(tile, options) {
 }
 
 // normalize tile headers
-export function normalizeTileHeaders(tileset) {
+export async function normalizeTileHeaders(tileset: Tileset3D) {
```
Comment (ibgreen) [inline modules/3d-tiles/src/lib/parsers/parse-3d-tile-header.ts:73]: Return type

File: modules/3d-tiles/src/lib/parsers/parse-3d-tile-header.ts [hunk 7]
```diff
@@ -94,13 +111,12 @@ export function normalizeTileHeaders(tileset) {
  * TODO Check if Tile3D class can be a return type here.
  * @param tileset
  */
-export async function normalizeImplicitTileHeaders(tileset: Tileset3D) {
-  if (!tileset.root) {
-    return null;
-  }
-
+export async function normalizeImplicitTileHeaders(
```
Comment (ibgreen) [inline modules/3d-tiles/src/lib/parsers/parse-3d-tile-header.ts:114]: Return types seem to be missing everywhere, maybe make a pass in a separate PR?

File: modules/3d-tiles/src/lib/parsers/parse-3d-tile-header.ts [hunk 8]
```diff
@@ -164,3 +180,7 @@ export async function normalizeImplicitTileData(tile, rootSubtree: Subtree, opti
 
   return tile;
 }
+
+function getImplicitTilingExtensionData(tile: Tile3D | null): ImplicitTilingExtension {
```
Comment (ibgreen) [inline modules/3d-tiles/src/lib/parsers/parse-3d-tile-header.ts:184]: JSDoc

**Review by ibgreen — APPROVED** [review]

File: modules/tile-converter/src/i3s-converter/helpers/gltf-attributes.ts (hunk 1 above)
Comment (dryabinin-actionengine) [reply modules/tile-converter/src/i3s-converter/helpers/gltf-attributes.ts]: Updated comment. This function need to copy arrayBuffers to avoid data corruption in worker.

File: modules/tile-converter/src/i3s-attributes-worker.ts (hunk 2 above)
Comment (dryabinin-actionengine) [reply modules/tile-converter/src/i3s-attributes-worker.ts]: Done

File: modules/3d-tiles/src/lib/parsers/helpers/parse-3d-tile-subtree.ts (hunk 3 above)
Comment (dryabinin-actionengine) [reply modules/3d-tiles/src/lib/parsers/helpers/parse-3d-tile-subtree.ts]: done

File: modules/3d-tiles/src/lib/parsers/helpers/parse-3d-tile-subtree.ts (hunk 4 above)
Comment (dryabinin-actionengine) [reply modules/3d-tiles/src/lib/parsers/helpers/parse-3d-tile-subtree.ts]: done

File: modules/3d-tiles/src/lib/parsers/helpers/parse-3d-tile-subtree.ts (hunk 5 above)
Comment (dryabinin-actionengine) [reply modules/3d-tiles/src/lib/parsers/helpers/parse-3d-tile-subtree.ts]: Done

File: modules/3d-tiles/src/lib/parsers/parse-3d-tile-header.ts (hunk 6 above)
Comment (dryabinin-actionengine) [reply modules/3d-tiles/src/lib/parsers/parse-3d-tile-header.ts]: Done

File: modules/3d-tiles/src/lib/parsers/parse-3d-tile-header.ts (hunk 7 above)
Comment (dryabinin-actionengine) [reply modules/3d-tiles/src/lib/parsers/parse-3d-tile-header.ts:114]: Yes, will create another PR

File: modules/3d-tiles/src/lib/parsers/parse-3d-tile-header.ts (hunk 8 above)
Comment (dryabinin-actionengine) [reply modules/3d-tiles/src/lib/parsers/parse-3d-tile-header.ts:191]: Done

File: modules/3d-tiles/src/lib/parsers/helpers/parse-3d-implicit-tiles.ts [hunk 9]
```diff
@@ -211,6 +211,8 @@ function formatTileData(
     type: getTileType(tile),
     lodMetricType,
     lodMetricValue,
+    geometricError: lodMetricValue,
+    transform: tile.transform,
     boundingVolume
     // Temp debug values. Remove when real implicit tileset will be tested.
     // x: tile.x,
```
Comment (belom88) [inline modules/3d-tiles/src/lib/parsers/helpers/parse-3d-implicit-tiles.ts:218]: I guess this commented code can be removed

**Review by belom88 — APPROVED** [review]
